generative_data_prep/processors/article_tokenizer.py

```diff
@@ -141,6 +148,200 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
+        will contain a single tokenized prompt completion pair from the original jsonl.
+        If self.packing_boundary is BoundaryType.JSONL, then each tokenized article will contain
+        the tokenized text from all the prompt completion pairs in the jsonl.
+
+        Args:
+            jsonl (List[dict]): A list of dictionaries where each dictionary represents a JSONL data item.
+            Each dictionary should contain 'prompt' and 'completion' keys.
+            The loaded input jsonl, in the form of {"prompt":"...", "completion":"..."}
+            or [{"prompt":"...", "completion":"..."}, {"prompt":"...", "completion":"..."}, ...]
+
+        Returns:
+            List[TokenizedArticle]: Tokenized articles that represent input jsonl.
+            A list of dictionaries where each dictionary represents a tokenized article.
+        """
+        if isinstance(jsonl, dict):
+            jsonl = [jsonl]
+
+        converted_jsonl = []
+        prompts = []
+        completions = []
+
+        """
+        TODO: Add a trick here to get the generation_prompt in tokenizer
+        This is a temporary solution, and the tokenizer should be modified in the future.
+        The tokenizer should be able to get the generation_prompt from the input text.
+        self.tokenizer.chat_template = "{% for message in messages %} {{'<|im_start|>' + message['role'] + '\n' +
+            message['content'] + '<|im_end|>' + '\n'}}{% endfor %}"
+        """
+
+        """
+        The mechanism designed here is to remember the original prompt and completion,
+        despite the transformation of the [prompt, completion] pair into a raw string
+          format with additional instruction tokens.
+
+        The following steps are performed:
+        1. Recognizes and preserves the original dialogue data format. (<prompt_placeholder>, <completion_placeholder>)
+        2. Converts the original [prompt, completion] format to the standard conversation [role, content] format.
+            {"prompt": "Hello, how are you?", "completion": "I'm doing great"} ->
+                {"role": "user", "content": "Hello, how are you?"}, {"role": "assistant", "content": "I'm doing great."}
+        """
+        for prompt_completion in jsonl:
+            if self.prompt_keyword not in prompt_completion:
+                err_msg = f"Prompt keyword required in every jsonl, {self.prompt_keyword} not found"
+                raise json.JSONDecodeError(err_msg, str(jsonl), 0)
+
+            if self.completion_keyword not in prompt_completion:
+                err_msg = f"Completion keyword required in every jsonl, {self.completion_keyword} not found"
+                raise json.JSONDecodeError(err_msg, str(jsonl), 0)
+
+            prompt = prompt_completion[self.prompt_keyword]
+            completion = prompt_completion[self.completion_keyword]
+            converted_jsonl.append({"role": "user", "content": DEFAULT_PROMPT_PLACEHOLDER})
```

Comment:
Do we want to hard code "user"/"assistant" here or do we want to allow users to define these? Also, do we want to consider the case when user wants to pass in system prompt?

Reply:
thanks for pointing this out, the format is hard coded here only according to hugging face version.